### ISSUE TITLE
How to Create the DataTemplate with Converter using StringBuilder(CodeBehind) in UWP ?

### ISSUE THREAD

**Issue by vijayarasan** [microsoft/microsoft-ui-xaml#4774]
I have tried to create the application with full code behind in UWP. But I have to get an exception. When I try to use XAML code it is working properly but the same code using StringBuilder Issue occurs.

Here is the working Xaml code:
```
<Page.Resources>
        <local:ColumnNameToColumnIndexConverter x:Key="ColumnNameToColumnIndexConverter"/>
    </Page.Resources>
    <Grid x:Name="parent">
        <Button>
            <Button.ContentTemplate>
                <DataTemplate>
                    <StackPanel>
                        <TextBox Width = '200' Height='50' Text = '{Binding Converter={StaticResource ColumnNameToColumnIndexConverter }}'  />
                    </StackPanel>
                </DataTemplate>
            </Button.ContentTemplate>
        </Button>
    </Grid>
```

Screenshot for your reference,
![WorkingXAMLCodeScreenShot](https://user-images.githubusercontent.com/58214164/114158109-f8248900-9941-11eb-8c86-7df26a24c096.png)

Here is my attempt to recreate it in CodeBehind, but it is unfortunately not working.

```
 //Create the DataTemplate in Code behind
            StringBuilder sb1 = new StringBuilder();
            sb1.Append("<DataTemplate xmlns='http://schemas.microsoft.com/winfx/2006/xaml/presentation\'>");
            sb1.Append("<StackPanel >");
            sb1.Append("<TextBox Width = '200' Height='50' Text = '{Binding  ElementName = \"btnDisplay\", Converter={StaticResource (ColumnNameToColumnIndexConverter)this.Resources[\"ColumnNameToColumnIndexConverter\"] }}'  />");
            sb1.Append("</StackPanel >");
            sb1.Append("</DataTemplate >");
            var output = (DataTemplate)XamlReader.Load(sb1.ToString());

            Button button = new Button();
            button.Name = "btnDisplay";
            button.Width = 150;
            button.Height = 50;
            button.ContentTemplate = output;
            parent.Children.Add(button);
```
Screenshot for your reference,
![CodeBehindConverter](https://user-images.githubusercontent.com/58214164/114157405-3a999600-9941-11eb-9ecc-ee4bc1c04629.png)
In above mentioned C# code to create the Code behind in UWP but I have to get an exception
I have attached a sample for your reference.

**Sample Link:** [StringBuilderBinding.zip](https://github.com/microsoft/microsoft-ui-xaml/files/6284580/StringBuilderBinding.zip)

Can you provide how to Create the DataTemplate code behind the UWP platform?

**Note:**
**Please refer to the MSDN link:** https://developercommunity.visualstudio.com/t/how-to-create-the-datatemplate-with-converter-usin/1381994

Regards,
Vijayarasan S

**Comment by StephenLPeters** [comment]
The string being produced by string builder doens't look equivalent to the xaml to me.  The string builder version has an element name property set to something which doesn't seem to exist?

**Comment by vijayarasan** [comment]
> The string being produced by string builder doens't look equivalent to the xaml to me. The string builder version has an element name property set to something which doesn't seem to exist?

Sorry for the delay.

We are a little unclear with your answer. Our provided XAML code used in MainWindow is working properly but the same code using StringBuilder Issue occurs. If possible please provide the properly working code in the string builder.

Regards,
Vijayarasan S

**Comment by StephenLPeters** [comment]
```
 //Create the DataTemplate in Code behind
            StringBuilder sb1 = new StringBuilder();
            sb1.Append("<DataTemplate xmlns='http://schemas.microsoft.com/winfx/2006/xaml/presentation\'>");
            sb1.Append("<StackPanel >");
            sb1.Append("<TextBox Width = '200' Height='50' Text = '{Binding  ElementName = \"btnDisplay\", Converter={StaticResource (ColumnNameToColumnIndexConverter)this.Resources[\"ColumnNameToColumnIndexConverter\"] }}'  />");
            sb1.Append("</StackPanel >");
            sb1.Append("</DataTemplate >");
            var output = (DataTemplate)XamlReader.Load(sb1.ToString());

            Button button = new Button();
            button.Name = "btnDisplay";
            button.Width = 150;
            button.Height = 50;
            button.ContentTemplate = output;
            parent.Children.Add(button);
```
This part of your string builder looks strange to me:
```
ElementName = \"btnDisplay\"
```
I think it can just be deleted. The DataTemplate will have as its binding context the object it is templating, so you shouldn't need to point it at your button. Additionally, I don't think you can simply name a UIElement and then sucessfully reference them from a markup string. In markup the named element and the reference to it need to exist in the same context, so I don't think the string you've built knows about the button you set the name on.

**Comment by vijayarasan** [comment]
Hi @StephenLPeters ,

Can you please provide any working solution to achieve the DataTemplate with Converter using StringBuilder in code behind? if possible please provide the working sample to achieve this.

Regards,
Vijayarasan S

**Comment by chrisglein** [comment]
Can we zoom out a bit and focus on what problem you're trying to solve? Constructing a Binding from code using DataTemplate markup also in code is a less common scenario (given that in code you can hook events and set up things directly without needing all the features of bindings, especially converters).

> I have tried to create the application with full code behind in UWP. 

Side note, this is an interesting scenario that has merit. We've done some hackathon-style incubations around it. What we found is that the further you go down this code-centric mindset the less you want to lean on markup-focused utilities like VisualStateManager and Bindings. What's your core motivation for building your app this way?

**Comment by vijayarasan** [comment]
Hi @chrisglein,

Sorry for the delay.

We have already created some applications with full code behind. So, we would like to make this application also code behind while using DataTemplate.

Regards,
Vijayarasan S

**Comment by RealTommyKlein** [comment]
I'm not familiar with the inner workings of StaticResources, but I'm guessing the StaticResource needs to be resolvable when the DataTemplate is parsed.  In the XamlReader.Load case, it can't find `ColumnNameToColumnIndexConverter` since it isn't declared in the scope.  Declaring an instance of the converter in the template itself worked for me - here's the updated code to construct the template, everything else is unchanged:

```
            //Create the DataTemplate in Code behind
            StringBuilder sb1 = new StringBuilder();
            sb1.AppendLine("<DataTemplate ");
            sb1.AppendLine("xmlns='http://schemas.microsoft.com/winfx/2006/xaml/presentation'");
            sb1.AppendLine("xmlns:x='http://schemas.microsoft.com/winfx/2006/xaml'");
            sb1.AppendLine("xmlns:local='using:StringBuilderBinding'");
            sb1.AppendLine(">");

            sb1.AppendLine("<StackPanel>");

            // Create the instance of the converter for the template to use
            sb1.AppendLine("<StackPanel.Resources>");
            sb1.AppendLine("<local:ColumnNameToColumnIndexConverter x:Key='ColumnNameToColumnIndexConverter'/>");
            sb1.AppendLine("</StackPanel.Resources>");

            sb1.AppendLine("<TextBox Width = '200' Height='50' Text = '{Binding  ElementName = \"btnDisplay\", Converter={StaticResource ColumnNameToColumnIndexConverter}}'  />");
            sb1.AppendLine("</StackPanel >");
            sb1.AppendLine("</DataTemplate >");
            var output = (DataTemplate)XamlReader.Load(sb1.ToString());
```

**Comment by AndrewKeepCoding** [comment]
I tried the sample code posted by @RealTommyKlein but the converter couldn't be resolved (XamlParseException) on v1.4.240211001.

To make it work, I needed to declare the converter in a XAML file that should be run by the XAML compiler. For example, a ResourceDictionary:

```xaml
<ResourceDictionary
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:StringBuilderBinding">

    <local:ColumnNameToColumnIndexConverter x:Key="ColumnNameToColumnIndexConverter" />

</ResourceDictionary>
```

This seems to be related to #7206.
